src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -2181,32 +2420,37 @@ public static TSource Max<TSource>(this IEnumerable<TSource> source)
             TSource value = default(TSource);
             if (value == null)
             {
-                foreach (TSource x in source)
+                using (IEnumerator<TSource> e = source.GetEnumerator())
                 {
-                    if (x != null && (value == null || comparer.Compare(x, value) > 0))
-                        value = x;
+                    if (!e.MoveNext()) return value;
+                    value = e.Current;
+                    // A do…while may seem more natural here, but profiled worse across a wide set of test data
```

Comment:
I'm surprised it was measurable in this case.  In addition to the two interface calls for every element, we also have the interface call to the comparer now for each element below.  When you say it profiled worse, was that just in the extreme case where things are dominated by nulls in the input sequence?  I wouldn't optimize for that case.